### ISSUE TITLE
Typescript typing trouble for immer proxy

### ISSUE THREAD

**Issue by clawoflight** [react-spring/zustand#98]
I'm a TS user trying to properly type an immer middleware, but I'm failing miserably:

Vanilla:
```javascript
// Turn the set method into an immer proxy
const immer = config => (set, get, api) => config(fn => set(produce(fn)), get, api)
```
TS:
```typescript
function immerConverter<T>(config: StateCreator<T>): StateCreator<T> {
  return (set: SetState<T>, get, api) =>
    config(fn => set(produce(fn)), get, api);
}
```
Leading to the following error:
```
Error:(46, 30) TS2345: Argument of type 'PartialState<T>' is not assignable to parameter of type '(...args: any[]) => any'.
  Type 'Partial<T>' is not assignable to type '(...args: any[]) => any'.
    Type 'Partial<T>' provides no match for the signature '(...args: any[]): any'.
```

Any ideas?

**Comment by mongris** [comment]
I did mine like:

```ts
function immer<T>(
  config: ((set: ((fn: (partial: T) => void) => void), get: GetState<T>, api: StoreApi<T>) => T)
) {
  return (set: SetState<T>, get: GetState<T>, api: StoreApi<T>) =>
    config((fn: any) => set(produce(fn)), get, api)
}
```

**Comment by clawoflight** [comment]
@mongris that turns off type checking when implementing functions inside create() though, and autocompletion along with it.
That is unacceptable.

@drcmda @JeremyRH can we please get fixed typings for immer usage? #96 is related.

**Comment by dai-shi** [comment]
@clawoflight @mongris Please continue the discussion in #96.
